service/http/http.go

```diff
@@ -299,3 +348,78 @@ type Component interface {
 	// will receive a request to just `/metrics`.
 	Handler() http.Handler
 }
+
+// lazyListener is a [net.Listener] which lazily initializes the underlying
+// listener.
```

Comment:
Should we put it in its own file?

Reply:
I'm not sure it's worth it, it's only 70 lines of code.